src/NECompletion-Morphic/CompletionEngine.class.st

```diff
@@ -257,6 +259,34 @@ CompletionEngine >> openMenu [
 		ifTrue: [ ^self closeMenu ].
 ]
 
+{ #category : #replacement }
+CompletionEngine >> replaceTokenInEditorWith: aString [
+	"Main API with the completion context.
+	Replace the current completion token (as it was extracted by the completion context) by aString.
+	aString may contain a keyword selector string where each keyword is separated by two spaces (between:  and:)
+	After replacing, set the caret after the first keyword.
+	The completion context uses this API to insert text into the text editor"
+
+	| caret old positionAfterFirstKeyword offset firstKeyword |
+	"Try to correctly replace the current token in the word.
+	The code editor should be able to do it by himself"
+	caret := self editor caret.
+	self editor
+		selectInvisiblyFrom: caret - context completionToken size
+		to: caret - 1.
+	old := self editor selection.
+	self editor replaceSelectionWith: aString.
+	
+	offset := NECPreferences spaceAfterCompletion
+		ifTrue: [ 1 ]
+		ifFalse: [ 0 ].
+	firstKeyword := (aString copyUpTo: $ ) size.
+	positionAfterFirstKeyword := caret + firstKeyword + offset - old size.
+	self editor selectAt: positionAfterFirstKeyword.
+	
+	self editor morph invalidRect: self editor morph bounds
+]
+
```

Comment:
The insertion of the text in the editor was moved from the menu morph to the completion engine, where it can decide how to do it properly.